cmd/notify-mailer/main.go

```diff
@@ -231,87 +254,138 @@ func emailsForReg(id int, dbMap dbSelector) ([]string, error) {
+		}
+		extraColNames = append(extraColNames, colName)
 	}
 
-	results := []recipient{}
+	var recipients []recipient
 	for {
-		record, err := reader.Read()
-		if err == io.EOF {
-			if len(results) == 0 {
-				return nil, fmt.Errorf("no entries after the header in CSV")
-			}
-			return results, nil
-		}
+		entry, err := recipientList.Read()
 		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Finished parsing the file.
+				break
+			}
 			return nil, err
 		}
-		if len(record) == 0 {
-			return nil, fmt.Errorf("empty line in CSV")
-		}
-		if len(record) != len(columnNames)+1 {
-			return nil, fmt.Errorf("Number of columns in CSV line didn't match header columns."+
-				" Got %d, expected %d. Line: %v", len(record), len(columnNames)+1, record)
-		}
-		id, err := strconv.Atoi(record[0])
+
+		// Attempt to coerce the value provided for `id` to a valid registration
+		// ID (int64).
+		entryID := entry[0]
+		id, err := strconv.ParseInt(entryID, 10, 64)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf(
+				"%q couldn't be parsed as a registration ID: %s", entryID, err)
 		}
-		recip := recipient{
-			id:    id,
-			Extra: make(map[string]string),
+
+		// Ensure our file doesn't contain duplicate entries for the same
+		// registration ID.
+		for _, recipient := range recipients {
+			if id == recipient.id {
+				return nil, fmt.Errorf("id: %d is duplicated", id)
+			}
 		}
-		for i, v := range record[1:] {
-			recip.Extra[columnNames[i]] = v
+
+		// Create a map of `Extra` columns (anything after `id`) to values.
+		extra := make(map[string]string)
+		for extraColNamesPos, value := range entry[1:] {
+			colValue := strings.TrimSpace(value)
+
+			// Ensure the column value isn't just a trailing delimiter.
+			if len(colValue) == 0 {
+				return nil, fmt.Errorf(
+					"line: %q contains duplicate or trailing delimiter", entry)
+			}
+			extra[extraColNames[extraColNamesPos]] = colValue
 		}
-		results = append(results, recip)
+
+		// Parsing of the row is complete.
+		recipients = append(recipients, recipient{id, extra})
 	}
+
+	// Ensure the supplied recipient list isn't empty.
+	if len(recipients) == 0 {
+		return nil, errors.New("recipient list contained 0 entries after the header")
+	}
+	return recipients, nil
+}
+
+// makeRecipientList determines the format of the recipient list file,
+// constructs a reader, and returns the result of parsing with that reader.
+func makeRecipientList(csvFilename, tsvFilename *string) ([]recipient, error) {
+	var filename string
+	var delimiter rune
+
+	if csvFilename != nil {
+		filename = *csvFilename
+	} else if tsvFilename != nil {
+		// We can use a `*csv.Reader` to parse a TSV file, we just have to
+		// supply a custom delimiter in rune form.
+		filename = *tsvFilename
+		delimiter = '\t'
+	}
+
+	contents, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	reader := csv.NewReader(contents)
+	if delimiter != 0 {
```

Comment:
Why use 0 as a special value here rather than unconditionally setting to `,` in the CSV path?